elkscmd/mtools/Mcopy.1
.TH MCOPY 1 local
.SH NAME
mcopy \- copy MSDOS files to/from Unix
.SH SYNOPSIS
.B mcopy
[
.B -tnvm
] unixfile a:msdosfile
.PP
.B mcopy
[
.B -tnvm
] unixfile [ unixfiles... ] a:msdosdirectory
.PP
.B mcopy
[
.B -tnm
] a:msdosfile unixfile
.PP
.B mcopy
[
.B -tnm
] a:msdosfile [ a:msdosfiles... ] unixdirectory
.SH DESCRIPTION
In the first two forms,
.I mcopy
copies the specified Unix file to the named MSDOS file or copies
multiple Unix files to the named MSDOS directory.  The last two forms
copy MSDOS files to Unix.  MSDOS sub directories are supported with
either the '/' or '\e\' separator.  The use of the '\e\' separator or
wildcards will require the names to be enclosed in quotes to protect
them from the shell.
.PP
The use of the 'a:' drive designation on the MSDOS files determines the
direction of the transfer.  A missing drive designation implies a Unix
file whose path starts in the current directory
.PP
.I Mcopy
will allow the following command line options:
.TP
.B t
Text file transfer.
.I Mcopy
will translate incoming carriage return/line feeds to line feeds.
.TP
.B n
No warning.
.I Mcopy
will not warn the user when overwriting an existing file.
.TP
.B v
verbose mode.
.TP
.B m
Preserve the file modification time.
.PP
If the target file already exists, and the
.I -n
option is not in effect,
.I mcopy
asks whether or not to overwrite the file.
.PP
The environmental variable MCWD may be used to establish a current
working directory (relative to the MSDOS diskette).
.SH SEE ALSO
mdir(1), mtype(1), mread(1), mwrite(1)
.SH BUGS
Wildcards are only applied to filenames, not to directory names.
Unlike MSDOS, the destination directory may not be omitted.


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/mtools/mcopy.c
/*
 * A front-end to the mread/mwrite commands.
 *
 * Emmet P. Gray			US Army, HQ III Corps & Fort Hood
 * ...!uunet!uiucuxc!fthood!egray	Attn: AFZF-DE-ENV
 * 					Directorate of Engineering & Housing
 * 					Environmental Management Office
 * 					Fort Hood, TX 76544-5057
 */

#include <stdio.h>

#define NONE	0
#define MREAD	1
#define MWRITE	2

main(argc, argv)
int argc;
char *argv[];
{
	extern int optind;
	extern char *optarg;
	int i, oops, msdos_args, unix_args, destination;
	char **nargv, *strcpy();
	void exit();
					/* get command line options */
	msdos_args = 0;
	unix_args = 0;
	oops = 0;
	while ((i = getopt(argc, argv, "tnvm")) != EOF) {
		switch (i) {
			case 't':
			case 'n':
			case 'v':
			case 'm':
				break;
			default:
				oops = 1;
				break;
		}
	}

	if (oops || (argc - optind) < 2) {
		fprintf(stderr, "Usage: mcopy [-tnvm] a:msdosfile unixfile\n");
		fprintf(stderr, "       mcopy [-tnvm] a:msdosfile [a:msdosfiles...] unixdirectory\n");
		fprintf(stderr, "       mcopy [-tnm] unixfile a:msdosfile\n");
		fprintf(stderr, "       mcopy [-tnm] unixfile [unixfiles...] a:msdosdirectory\n");
		exit(1);
	}
					/* what is the destination drive? */
	destination = NONE;
					/* first file */
	if (argv[optind][1] == ':') {
		if (argv[optind][0] == 'a' || argv[optind][0] == 'A')
			destination = MREAD;
	}
					/* last file */
	if (destination == NONE && argv[argc-1][1] == ':') {
		if (argv[argc-1][0] == 'a' || argv[argc-1][0] == 'A')
			destination = MWRITE;
	}
	if (destination == NONE) {
		fprintf(stderr, "mcopy: no 'a:' designation specified\n");
		exit(1);
	}
					/* strip out the fake "drive code" */
	for (i=optind; i<argc; i++) {
		if (argv[i][1] == ':') {
			switch (argv[i][0]) {
				case 'a':
				case 'A':
					if (argv[i][2] == '\0')
						strcpy(argv[i], ".");
					else
						strcpy(argv[i], &argv[i][2]);
					msdos_args++;
					break;
				case 'c':
				case 'C':
					fprintf(stderr, "mcopy: 'c:' is not used to designate Unix files\n");
					exit(1);
				default:
					fprintf(stderr, "mcopy: unknown drive '%c:'\n", argv[i][0]);
					exit(1);
			}
			continue;
		}
					/* if no drive code, its a unix file */
		unix_args++;
	}
					/* sanity checking */
	if (!msdos_args || !unix_args) {
		fprintf(stderr, "mcopy: unresloved destination\n");
		exit(1);
	}
	if ((destination == MWRITE && msdos_args > 1) || (destination == MREAD && unix_args > 1)) {
		fprintf(stderr, "mcopy: duplicate destination files\n");
		exit(1);
	}
	/*
	 * Copy the *argv[] array in case your Unix doesn't end the array
	 * with a null when it passes it to main()
	 */
	nargv = (char **) malloc((argc+1) * sizeof(*argv));
	nargv[argc] = NULL;
	for (;--argc>=0;)
		nargv[argc] = argv[argc];

	if (destination == MWRITE) {
		nargv[0] = "mwrite";
		execvp("mwrite", nargv);
		perror("execvp: mwrite") ;
	}
	else {
		nargv[0] = "mread";
		execvp("mread", nargv);
		perror("execvp: mmread") ;
	}
}
